cexec.1
.TH cexec 1 "1.0" "Internet Connection, Inc." "Ad-Hoc Clusters"
.SH NAME
cexec \- submit work for a cluster application
.SH SYNOPSIS
.B cexec appid.pub [peer env]

Locates a copy of
.I cservice(1)
running with the private key for
.B appid.pub
and runs the (possibly) remote application otherwise transparently.
.SH DESCRIPTION
.B cexec
first locates a running copy of
.I cservice(1)
on the network called
.I $CEXEC_GROUP
or
.I $GROUP
and waits for a job-acceptance connection.
.B cexec
then ties all of the file descriptors up to
.I $CEXEC_MAXFD
or
.I $MAXFD
and multiplexes them over the connection to the running
.I cservice(1)
which is doing the same task for the application it is running.

.B cexec
can be thought of like
.I rexec(1)
except instead of specifying a host to run the task on,
.B cexec
will run the task on the least-loaded most-capable host on the network.

The
.B peer env
is a list of environment variables and their values (remember to quote!)
in
.I KEY=VALUE
format. These are prefaced with the string
.B PEER_
by
.I cservice(1)
before the application has a chance to see them.
.SH ENVIRONMENT
.B $CEXEC_GROUP
or
.B $GROUP
\- contains the "group name" - an IP address (possibly multicast) representing
the address to be used for sending announcements.
.I Default: 255.255.255.255

.B $CEXEC_MAXFD
or
.B $MAXFD
\- contains the highest-numbered file descriptor to scan when multiplexing
file descriptors over to the peer.
.I Default: 32

.B $CEXEC_TIMEOUT_ACCEPT
\- the maximum amount of time (in seconds) that
.B cexec
will wait when attempting to locate a running copy of
.I cservice(1)
\- set to 0 in order to wait forever.
.I Default: 60 seconds

.B $CEXEC_FD_FOR_ERRORS
\- a numeric file descriptor that must be greater than
.B $CEXEC_MAXFD
that extra status information will be written to after job completion or error.

.B $CEXEC_LISTEN_PORTS
\- a list of port numbers as
.I a,b,c
or including ranges
.I a:b,c:d
that will be the TCP port numbers
.B cexec
will expect a response on.
.SH DIAGNOSTIC
Setting
.B $CEXEC_RATOUT_TEST
will generate a ratout alert for
.I crat(1)
immediately after locating a host.

Setting
.B $CEXEC_TIMEOUT_PINGPONG
to a number of seconds will allow you to test your application for
what happens when the pingpong timeout expires. This shouldn't ordinarily
be used - even when you're experiencing other problems.
.SH EXIT CODES
.B cexec
normally maps all exit codes back to the calling process from the
remote side. The exit code ``111'' may be synthesized by
.B cexec
to announce its own errors. All of these errors should be treated as
temporary.
.SH SEE ALSO
ckeygen(1), cservice(1), crat(1), cproxy(1)


CHANGELOG
2008-08-07 18:02  geocar

	* Makefile, bio.c, build-cross.sh, cexec.c, ckeygen.c, cproxy.c,
	  crat.c, cservice.c, deploy.sh, hash.c, keyfile.c, load.c,
	  multipipe.c, netnum.c, netnum.h, parity.c, sha1.c, sha1.h,
	  strline.c, strline.h, types.h: port to 64bit system

2007-03-02 22:05  geocar

	* deploy.sh: amd64

2007-01-30 21:27  geocar

	* cexec.c: wee

2006-12-08 22:48  geocar

	* bio.c, cexec.c, cservice.c, deploy.sh: moving it around...

2006-12-08 17:48  geocar

	* bio.c, build-cross.sh, cexec.c, cio.c, cservice.c, deploy.sh: fix
	  base when doing copyout; dont close on write (sigpipe helps
	  noone) and sparc is dead. oh, and nul-group is all-group

2006-07-31 06:46  geocar

	* cservice.c: support $BOOST mode

2006-07-31 06:46  geocar

	* load.c: use gnu load

2006-06-23 06:59  geocar

	* cexec.c: dont regenerate parity (we cannot remember long enough)

2006-06-23 06:59  geocar

	* cservice.c: fix delays, and make pubkey toplevel so we can test

2006-06-23 05:27  geocar

	* Makefile: uses sockop

2006-06-23 05:27  geocar

	* cexec.c: add testing feature, and skipto to try next-best host

2006-06-23 05:26  geocar

	* cservice.c: check signature for safety and use sockop_nonblock

2006-06-23 05:26  geocar

	* hash.c: zero s1/s2 barrels

2006-06-23 05:26  geocar

	* keyfile.c: check zero length vlpoint

2006-06-23 05:26  geocar

	* cio.c, sockop.c: bleh, convenience i suppose

2005-11-08 07:51  geocar

	* cexec.c: remove did_io optimization; it's complicated

2005-03-17 10:07  geocar

	* ckeygen.1: adding documentation about -stdin flag

2005-03-17 10:06  geocar

	* ckeygen.c: generate keys based on stdin with -stdin option now...

2005-03-17 10:06  geocar

	* cexec.c: did_io when writing too...

2005-03-01 12:58  geocar

	* cservice.c: memalloc isn't fatal PEER_vars, don't attempt to dup2
	  a -1 fd (bitmap)

2005-03-01 11:31  geocar

	* cproxy.c: don't waste file descriptors all wily-nilly, also
	  multihomed SO_BINDTODEVICE capable boxes are now assisted...

2005-03-01 11:31  geocar

	* cexec.c: put lseek hack back in

2005-02-17 08:11  geocar

	* parity.c: use urandom instead of random; the lifetime of the
	  check-bits are so short that other attacks are more likely to be
	  useful

2005-02-17 07:59  geocar

	* cservice.c: do bit flipping if near or far is running on a really
	  strange platform

2005-02-17 07:58  geocar

	* cexec.c: return didio hack

2005-02-17 07:58  geocar

	* cexec.c, cluster.h: move bytesex constant out of cexec

2005-02-17 07:57  geocar

	* bio.c: sink output in blast and copy

2005-02-17 06:27  geocar

	* LICENSE, Makefile, cexec.c, cio.c, cio.h, cservice.c, keyfile.c:
	  moved cread/cwrite out of keyfile and changed write-hack to use
	  4bytes instead of a vlpoint. the write-hack might prove useful
	  someday...

2005-02-16 14:40  geocar

	* cexec.c, cservice.c: turn keepalive back on

2005-02-16 13:02  geocar

	* cexec.c, cservice.c: don't use keepalive anymore, protect
	  main-channel for sanity, and add an extra vlpoint-exchange just
	  to force data both ways on the socket before working/forking

2005-02-16 13:00  geocar

	* bio.c: protect fcntl changes

2005-02-16 12:50  geocar

	* bio.c: turn off nonblock when blasting, and honor EAGAIN on
	  bio_more

2005-02-16 12:47  geocar

	* keyfile.c: turn off nonblock when in cread/cwrite

2005-02-16 09:54  geocar

	* cservice.1, cservice.c: set $REMOTE_ADDR, $REMOTE_PORT,
	  $SERVER_ADDR, $SERVER_PORT so cservice child can do things like
	  ident-check the cexec peer if it likes

2005-02-16 09:53  geocar

	* crat.c: unknown keys are listed in hex for brevity

2005-02-16 09:53  geocar

	* strline.c, strline.h: support for hex numbers too

2005-02-16 08:29  geocar

	* LICENSE, Makefile, cexec.c, crat.c, cservice.c, fdm.c, pair.c,
	  sockop.c, sockop.h: move sockops into a separate file

2005-02-16 08:28  geocar

	* cproxy.c: bugfix; detect hangup on write; move expires operation
	  before gathering fds so cproxy doesn't bomb

2005-02-15 10:44  geocar

	* cexec.c: restart processing if we haven't done any I/O but still
	  got an error,

2005-02-15 10:44  geocar

	* bio.c: treat EINTR and EAGAIN the same

2005-02-15 10:43  geocar

	* cservice.c: map fd masks correctly; increased timeouts, don't try
	  and read from wronly fds, use PIPE_BUF for buffers, and send
	  close alerts for write-errors

2005-02-15 10:42  geocar

	* fdm.c: map rdonly file descriptors correctly now; treat ttys
	  specially

2005-02-15 10:41  geocar

	* keyfile.c: treat EINTR and EAGAIN the same; also be pendantic
	  about short != 16 bits

2005-02-14 17:09  geocar

	* cexec.c: use a different (better?) announcement schedule

2005-02-14 17:09  geocar

	* cproxy.c: un-nonblock fds, turn on keepalive/ndelay on both
	  sides, honor EAGAIN

2005-02-14 17:09  geocar

	* cservice.1, cservice.c: lower default concurrency, drop
	  unnecessary checks, and recover from autorat failures faster

2005-02-14 16:35  geocar

	* load.c: system_load() isn't required to actually function

2005-02-14 11:01  geocar

	* cexec.1, ckeygen.1, cproxy.1, crat.1, cservice.1: adding
	  documentation

2005-02-14 08:57  geocar

	* cexec.c: don't waste an fd

2005-02-14 08:47  geocar

	* cexec.c, cservice.c: added pingpong detection, configurable
	  timeouts, configurable port ranges, using simplexes instead of
	  duplexes where possible, sending fd usage map, and debugging
	  nonsense for cexec

2005-02-14 08:43  geocar

	* crat.c, strline.c, strline.h: move string opers stuff

2005-02-14 08:42  geocar

	* fdm.c: fixes in detecting file descriptors on Linux

2005-02-14 08:32  geocar

	* pair.c: turn in so_linger on our pair links.

2005-02-14 08:30  geocar

	* parity.c: use /dev/random if available

2004-12-16 07:21  geocar

	* cservice.c: moving waitpid chunk earlier,change autorat timeouts,
	  and disable need_delay after delay (procrastinating autorats)

2004-12-16 06:37  geocar

	* cservice.c: default maxchildren now 100

2004-12-16 06:31  geocar


cexec.c
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>
#include <stdlib.h>

#include "cluster.h"
#include "netnum.h"

void parity_init(void);
int parity(void);

#include "bio.h"
struct bio biodata;
struct bio copyout;

#include "ec_crypt.h"
#include "ec_vlong.h"

#ifndef AF_LOCAL
#define AF_LOCAL AF_UNIX
#endif

/* bottom 4 bytes are the host to rat out,
 * the top four are H(host,H(pubkey))
 *
 * this is a special message that's always
 * EXACTLY 8 characters long
 */
#include "strline.h"

static unsigned int timeout_accept = 60;
static unsigned int timeout_pingpong = 600;
static int maxfd;


static unsigned char rathost[9];
static void ratout(int bfd, in_addr_t group, int ee)
{
	struct sockaddr_in sin;
	int fdout;

	if (getenv("CEXEC_FD_FOR_ERRORS")) {
		fdout = atoi(getenv("CEXEC_FD_FOR_ERRORS"));
		if (fdout > -1 && (fdout > maxfd || maxfd == -1)) {
			char outline[1024];
			unsigned char *op;
			unsigned int ip;
			unsigned int i;
			unsigned int len;
			int r;

			ip = rathost[0]
			| (rathost[1] << 8)
			| (rathost[2] << 16)
			| (rathost[3] << 24);

			op = outline;

			if (!ip) {
				op += static_str(op, "[cluster]");
			} else {
				op += ip_print(op, ip);
			}
			if (ee == -1 || ee == -5) {
				op += static_str(op, " failed to respond after ");
				if (ee == -1) {
					op += number_print(op, timeout_pingpong);
				} else {
					op += number_print(op, timeout_accept);
				}
				op += static_str(op, " seconds\n");
			} else if (ee == -2) {
				op += static_str(op, " disconnected unexpectedly\n");
			} else if (ee == -3) {
				op += static_str(op, " violated protocol\n");
			} else if (ee == -4) {
				op += static_str(op, " test alert\n");
			} else if (ee == -6) {
				op += static_str(op, " could not read signature\n");
			} else if (ee == -7) {
				op += static_str(op, " signature didn't verify\n");
			} else if (ee < 0) {
				op += static_str(op, " unknown error\n");
			} else {
				op += static_str(op, " exited with code ");
				op += number_print(op, (unsigned int)ee);
				op += static_str(op, "\n");
			}

			len = op - ((unsigned char *)outline);
			for (i = 0; i < len;) {
				do {
					r = write(fdout, outline+i, len-i);
				} while (r == -1 && errno == EINTR);
				if (r < 1) break; /* already error cond. */
				i += r;
			}
		}
	}

	/* attempt to rat out rathost */
	if (rathost[0] || rathost[1] || rathost[2] || rathost[3]) {
		memset(&sin, 0, sizeof(sin));
		sin.sin_family = AF_INET;
		sin.sin_port = htons(CLUSTER_PORT);
		memcpy(&sin.sin_addr, &group, sizeof(group));
		if (ee < 0) {
			(void)sendto(bfd, rathost, 8, 0,
				      (struct sockaddr *)&sin, sizeof(sin));
		} else {
			rathost[8] = (ee & 0xFF);
			(void)sendto(bfd, rathost, 9, 0,
				      (struct sockaddr *)&sin, sizeof(sin));
		}
	}

	if (ee < 0) {
		exit(111);
	}
	exit(ee);
}

static int try_bind(int portno)
{
	struct sockaddr_in sin;
	int i, sfd;

	sfd = socket(AF_INET, SOCK_STREAM, 0);
	if (sfd == -1) return -1;

	memset(&sin, 0, sizeof(sin));
	sin.sin_family = AF_INET;
	sin.sin_addr.s_addr = INADDR_ANY;

	sin.sin_port = htons(portno);
	sockop_keepalive(sfd, 1);
	sockop_ndelay(sfd, 0);
	sockop_reuseaddr(sfd, 1);
	if (bind(sfd, (struct sockaddr *)&sin, sizeof(sin)) == -1) {
		(void)close(sfd);
		return -1;
	}
	if (listen(sfd, 1) == -1) {
		(void)close(sfd);
		return -1;
	}
	return sfd;
}


int main(int argc, char *argv[])
{
	struct sockaddr_in sin;
	unsigned char header[4];
	unsigned int bohack;
	unsigned char rbuf[512], *rp, *hp, *pp;
	unsigned char cbuf[CLUSTER_SYNC];
	unsigned int len, subfd;
	unsigned char *hashbuf;
	unsigned int hashbuf_size;
	int start_port, end_port;
	in_addr_t ina, thost;
	int use_thost;
	int i, j, r, bfd, sfd, fd, nm;
	struct timeval tv;
	time_t now, start;
	fd_set rfds, use;
	fd_set lr, lw;
	vl_point pub, s1, s2;
	cp_pair sig;
	int skip_send;
	char *avoidh;
	int avoidc;
	char *q, *p;

	maxfd = -1;
RESTART:
	if (maxfd > -1) {
		for (i = 0; i <= maxfd; i++) {
			(void)lseek(i, 0, SEEK_SET);
		}
	}

	gf_init();
	memset(rathost, 0, sizeof(rathost));
	signal(SIGPIPE, SIG_IGN);

	if (argc < 2) {
#define M "Usage: cexec app [env...]\n"
		for (i = 0; i < sizeof(M)-1;) {
			do {
				r = write(2, M+i,(sizeof(M)-1)-i);
			} while (r == -1 && errno == EINTR);
			if (r < 1) exit(255);
			i += r;
		}
		exit(0);
#undef M
	}

	fd = open(argv[1], O_RDONLY);
	if (fd == -1) {
		q = (char*)malloc(strlen(argv[1])+5);
		if (!q) exit(111);
		strcpy(q, argv[1]);
		strcat(q, ".pub");
		fd = open(q, O_RDONLY);
		if (fd == -1) exit(111);
		(void)free(q);
	}
	if (!vlpoint_read(pub, fd)) exit(111);
	(void)close(fd);

	if (!bio_setup_malloc(&biodata, PIPE_BUF)) exit(111);
	if (!bio_setup_malloc(&copyout, PIPE_BUF)) exit(111);

	q = (char *)getenv("CEXEC_TIMEOUT_ACCEPT");
	if (q) timeout_accept = atoi(q);
	if (timeout_accept < 0) timeout_accept = 0;

	/* shouldn't set this... */
	q = (char *)getenv("CEXEC_TIMEOUT_PINGPONG");
	if (q) timeout_pingpong = atoi(q);
	if (timeout_pingpong < 120) timeout_pingpong = 120;

	maxfd = -1;
	q = (char *)getenv("CEXEC_MAXFD");
	if (!q) q = (char *)getenv("MAXFD");
	if (q) maxfd = atoi(q);
	if (maxfd < 0) maxfd = 32;
	maxfd = fdmap(maxfd+1, &lr, &lw);
	if (maxfd >= 0xFF00) exit(111); /* not allowed */

	parity_init();

	bfd = socket(AF_INET, SOCK_DGRAM, 0);
	if (bfd == -1) exit(111);

	q = (char *)getenv("CEXEC_GROUP");
	if (!q) q = (char *)getenv("GROUP");
	if (!q) q = "255.255.255.255";
	ina = inet_addr(q);

	i = *((unsigned char *)&ina);
	if (i >= 224 && i < 255) {
		/* multicast- but don't subscribe */
		i = 1; (void)setsockopt(bfd, IPPROTO_IP, IP_MULTICAST_LOOP, &i, sizeof(i));
		i = 31; (void)setsockopt(bfd, IPPROTO_IP, IP_MULTICAST_TTL, &i, sizeof(i));
	}
	sockop_broadcast(bfd, 1);
	sockop_reuseaddr(bfd, 1);

	/* quorom */
	q = (char *)getenv("CEXEC_REJECT_SERVER");
	if (!q) q = (char *)getenv("CEXEC_REJECT_SERVERS");
	avoidc = 0;
	if (q) {
		avoidh = (char*)malloc(strlen(q)+1);
		while (q) {
			p = strchr(q, ' ');
			if (p) *p = 0;
			thost = inet_addr(q);
			if (p) { *p=' '; q = p + 1; } else q = 0;
			memcpy(avoidh+(avoidc*4), &thost, 4);
			avoidc++;
		}
	} else {
		avoidh = 0;
	}

	/* good for testing... */
	q = (char *)getenv("CEXEC_ACCEPT_SERVER");
	if (q) {
		thost = inet_addr(q);
		use_thost = 1;
		timeout_accept = 0;
	} else {
		use_thost = 0;
	}

	q = (char *)getenv("CEXEC_LISTEN_PORTS");
	if (!q) q = (char *)getenv("CEXEC_LISTEN_PORT");
	if (!q) q = (char *)getenv("CEXEC_PORTS");
	if (!q) q = (char *)getenv("CEXEC_PORT");
	if (!q) {
		sfd = try_bind(0);
	} else {
		for (sfd = -1; sfd == -1;) {
			for (p = q; isdigit(((unsigned int)(*p))); p++);
			j = (unsigned char)*p;
			*p = 0; p++;
	
			start_port = atoi(q);
			if (start_port == -1) break;

			q = p;
			if (j == ':' || j == '-') {
				while (*q == ':' || *q == '-') q++;
				for (p = q; isdigit(((unsigned int)(*p))); p++);
				j = (unsigned char)*p;
				*p = 0; p++;
				end_port = atoi(q);
				if (end_port == -1) {
					sfd = try_bind(start_port);
					break;
				}

				while (start_port <= end_port) {
					sfd = try_bind(start_port);
					if (sfd != -1) break;
					start_port++;
				}
				q = p;
			}
			if (j == 0) {
				sfd = try_bind(start_port);
				break;
			}
			while (*q && !isdigit(((unsigned int)*q))) q++;
			if (!*q) break;
		}
	}
	if (sfd == -1) exit(111);

	i = sizeof(sin);
	if (getsockname(sfd, (struct sockaddr *)&sin, &i) == -1) exit(111);

	rp = rbuf + CLUSTER_SYNC;
	memcpy(rp, &sin.sin_port, sizeof(sin.sin_port));
	rp += sizeof(sin.sin_port);

	/* let the server know how many file descriptors we'll want */
	put_uint16(rp, maxfd+1); rp += 2;

	/* put hash of public key */
	hashbuf_size = vlpoint_size(pub);
	hashbuf = (char *)malloc(hashbuf_size+CLUSTER_SYNC);
	if (!hashbuf) exit(111);
	vlpoint_pack(pub, hashbuf+CLUSTER_SYNC, hashbuf_size);
	hp = rp; rp += 4;
	
	for (i = 2; i < argc; i++) {
		j = strlen(argv[i])+1;
		if (j+rp > rbuf+sizeof(rbuf)-2) exit(111);
		memcpy(rp, argv[i], j);
		rp += j;
	}
	if (rp + ((maxfd * 4) / 6) + 18 < rbuf+sizeof(rbuf)-2) {
		/* eh, it was hopeful */
		memcpy(rp, "CEXEC FD BITMAP=", 16);
		rp += 16;
		*rp = 0x80;
		for (i = j = 0; i <= maxfd; i++) {
			r = 0;
			if (FD_ISSET(i, &lr))
				r |= 1;
			if (FD_ISSET(i, &lw))
				r |= 2;
			*rp |= (r << j);
			j += 2;
			if (j > 5) {
				rp++;
				*rp = 0x80;
				j = 0;
			}
		}
		if (j != 0) rp++;
		*rp++ = 0; /* end of bitmap */

	}
	*rp++ = 0; /* null terminate strings */

	time(&start); memcpy(&now, &start, sizeof(time_t));
	skip_send = 0;

	/* write random challenge */
	for (i = 0, pp = rbuf; i < sizeof(cbuf); i++) {
		do {
			*pp = ((unsigned int)parity())
				| (((unsigned int)parity()) << 1)
				| (((unsigned int)parity()) << 2)
				| (((unsigned int)parity()) << 3)
				| (((unsigned int)parity()) << 4)
				| (((unsigned int)parity()) << 5)
				| (((unsigned int)parity()) << 6)
				| (((unsigned int)parity()) << 7);
		} while (!*pp);
		pp++;
	}

	for (fd = -1; fd == -1 && 
	(timeout_accept == 0 || now-start < timeout_accept); time(&now)) {

		/* update hash */
		memcpy(hashbuf, rbuf, CLUSTER_SYNC);
		put_uint32(hp, hash(hashbuf, hashbuf_size+CLUSTER_SYNC));

		memset(&sin, 0, sizeof(sin));
		sin.sin_family = AF_INET;
		memcpy(&sin.sin_addr, &ina, sizeof(in_addr_t));
		sin.sin_port = htons(CLUSTER_PORT);

		for (j = 1, r = 0; r == 0 && j < 32; j++) {
			if (!skip_send && sendto(bfd, rbuf, rp-rbuf, 0,
				      (struct sockaddr *)&sin, sizeof(sin)) == -1) {
				exit(111);
			}
	
			do {
				FD_ZERO(&rfds);
				FD_SET(sfd, &rfds);
				tv.tv_sec = (j >> 1) & 3; tv.tv_usec = 0;
/*				tv.tv_sec = (32-j) & 7; tv.tv_usec = 0; */
				r = select(sfd+1, &rfds, (fd_set *)0,
						(fd_set *)0, &tv);
			} while (r == -1 && errno == EINTR);
		}
		skip_send = 0;
		if (r == 0) continue; /* failed to get picked up */

		i = sizeof(sin);
		fd = accept(sfd, (struct sockaddr *)&sin, &i);
		if (fd == -1) continue;

		if (use_thost) {
			if (memcmp(&sin.sin_addr, &thost, 4) != 0) {
				(void)close(fd);
				skip_send = 1;
				fd = -1;
				continue;
			}
		}
		for (j = 0; j < avoidc; j += 4) {
			if (memcmp(&sin.sin_addr, avoidh+j, 4) != 0) {
				(void)close(fd);
				skip_send = 1;
				fd = -1;
				goto AVOIDED_HOST;
			}
		}

		memcpy(rathost, &sin.sin_addr, 4);
		put_uint32(rathost+4, hash(hashbuf+CLUSTER_SYNC, hashbuf_size));
		put_uint32(rathost+4, hash(rathost, 8));

		sockop_keepalive(fd, 1);
		sockop_ndelay(fd, 0);

		if (!vlpoint_read(sig.s, fd) || !vlpoint_read(sig.r, fd)) {
			(void)close(fd);
			if (use_thost) ratout(bfd, ina, -6);
			skip_send = 1;
			fd = -1;
			continue;
		}

		hash_point(s1, s2, rbuf, CLUSTER_SYNC);
		if (!cp_verify(pub, s2, &sig)) {
			/* failed verification */
			(void)close(fd);
			if (use_thost) ratout(bfd, ina, -7);
			skip_send = 1;
			fd = -1;
			continue;
		}

		bohack = CLUSTER_BYTESEX;
		memcpy(header, (void*)&bohack, 4);
		if (!cwrite(fd, header, 4)) {
			/* can't write? */
			(void)close(fd);
			skip_send = 1;
			fd = -1;
			continue;
		}
AVOIDED_HOST:
		do { /* nothing todo */ } while(0);
	}
	if (fd == -1) {
		ratout(bfd, ina,-5);

	} else if (getenv("CEXEC_RATOUT_TEST")) {
		ratout(bfd, ina,-4);
	}

	/* sin contains the host we're interested in */

	/* we have our worker bee */
	(void)close(sfd);

	FD_ZERO(&rfds);
	FD_SET(fd, &rfds); nm = fd;
	for (i = 0; i <= maxfd; i++) {
		if (FD_ISSET(i, &lr)) FD_SET(i, &rfds);
		if (i > nm) nm = i;
	}

	/* are we still here? */
	time(&start);
	for (;;) {
		fdset_copy(nm+1, &rfds, &use);
		do {
			tv.tv_sec = 5;
			tv.tv_usec = 0;
			r = select(nm+1, &use, (fd_set *)0, (fd_set *)0, &tv);
		} while (r == -1 && errno == EINTR);
		if (r == -1) {
			/* bad fd? maybe? */
			ratout(bfd, ina, -2);
			goto RESTART;
		}
		if (r == 0) {
			time(&now);
			/* 600 seconds with no contact? wow. */
			if (now - start > timeout_pingpong) {
				ratout(bfd,ina,-1);
				goto RESTART;
			}

			/* ping */
			put_uint16(header, 0xFFFF);
			put_uint16(header+2, 0xFFFF);
			bio_headout(&copyout, header, 4);
			if (!bio_blast(fd, &copyout, 0, 0)) {
				ratout(bfd,ina,-2);
				goto RESTART;
			}
			bio_base(&copyout);
			continue;
		}

		/* some activity someplace... */
		time(&start);
		if (FD_ISSET(fd, &use)) {
			do {
				/* demultiplex input */
				if (!bio_need(fd,&biodata,4)
				|| !bio_headin(&biodata, header, 4)) {
					ratout(bfd,ina,-2);
					goto RESTART;
				}
				len = get_uint16(header);
				subfd = get_uint16(header+2);

				if (len == 0xFFFF && subfd == 0xFFFF) {
					/* pong */
					continue;
				}

				if (subfd == 0) {
					/* the END! */
					ratout(bfd,ina,len);
				}
				subfd--; /* now an fd in map */
				if (subfd > maxfd) {
					ratout(bfd,ina,-3);
					goto RESTART;
				}
		
				/* close signal */
				if (len == 0) {
					if (subfd != fd) {
						FD_CLR(subfd, &rfds);
						FD_CLR(subfd, &lr);
						(void)close(subfd);
					}
				} else {
					if (!bio_copy(fd, subfd, &biodata,
								4, len)) {
						/* ack! */
						ratout(bfd,ina,-2);
						goto RESTART;
					}
				}
			} while (bio_base(&biodata));

			FD_CLR(fd, &use); /* in case fd < maxfd */
		}
		for (i = 0; i <= maxfd; i++) {
			if (!FD_ISSET(i, &lr)) continue;
			if (!FD_ISSET(i, &use)) continue;
			bio_skipin(&copyout, 4);
			if (!bio_more(i, &copyout)) {
				/* okay, channel closed.
				 * send to peer
				 */
				FD_CLR(i, &rfds);
				FD_CLR(i, &lr);
				(void)close(i);

				/* close command */
				put_uint16(header, 0);
			} else {
				j = bio_ready(&copyout) - 4;
				if (!j) continue;

				put_uint16(header, j);
			}

			put_uint16(header+2, i+1);
			bio_headout(&copyout, header, 4);
			if (!bio_blast(fd, &copyout, 0, 0)) {
				ratout(bfd,ina,-2);
				goto RESTART;
			}
			bio_base(&copyout);
		}
	}
	exit(127);
}
